acceptance/tests/agent/agent_can_reconnect_after_master_failure.rb

```diff
@@ -0,0 +1,57 @@
+require 'securerandom'
+test_name 'C97323: agents can reconnect after master crash/restart' do
+
+  username = 'er0ck'
+  teardown do
+    step 'stop puppet agent' do
+      on(agents, puppet_resource("service puppet ensure=stopped"))
+    end
+    step 'remove test user' do
+      on(agent, puppet_resource("user #{username} ensure=absent"))
+    end
+  end
+
+  environmentpath = master.puppet['environmentpath']
+  environment     = master.puppet['environment']
+  step 'create a file resource manifest for checking the puppet runs' do
+    manifest = "user { '#{username}': ensure=>present }"
+    apply_manifest_on(master, manifest, :catch_failures => true)
+    apply_manifest_on(master, <<MANIFEST, :catch_failures => true)
+     file {'#{environmentpath}/#{environment}/manifests/site.pp':
+       content => 'user { "#{username}": ensure=>present }'
+     }
+MANIFEST
+  end
+
+  step 'start puppet master' do
+    with_puppet_running_on(master, {}) do
+      agents.each do |agent|
+        step 'start puppet agent daemon at rapid runinterval' do
+          on(agent, puppet_resource("service puppet ensure=running"))
+        end
+        step "check users list for successful agent run" do
+          on(agent, puppet_resource("user #{username}")).stdout do |result|
+            assert_match(/#{username}/, result, 'did not find evidence of successful puppet run')
+          end
+        end
+        step 'kill master, clear user resource' do
+          on master, "service #{master['puppetservice']} stop"
+          on(agent, puppet_resource("user #{username} ensure=absent"))
```

Comment:
To test the behavior the agent needs to connect to the master process (completed TCP and SSL handshakes) and send an HTTP request to the server. If the server dies without replying with a TCP RST, then the agent will block indefinitely (before the fix). If you try to use beaker to stop the puppet master while the agent is running, it will most likely send a RST to the agent, invalidating the test. In that case, you'll probably get an EOF error described in PUP-3238.

It might be possible to `kill -9` the server process, but you'd have to do that sometime after the agents sends the HTTP request and before the server replies. I'm not sure how to do that in a beaker test?